argoverse/evaluation/eval_detection.py

```diff
@@ -0,0 +1,274 @@
+    gt_fpath: Path = Path("/data/argoverse")
+    fig_fpath: Path = Path("figs")
+    dt_cfg: DetectionCfg = DetectionCfg()
+
+    def evaluate(self) -> pd.DataFrame:
+        """Evaluate detection output and return metrics.
+
+        Returns:
+            The evaluation metrics.
+        """
+
+        dt_fpaths = list(self.dt_fpath.glob("*/per_sweep_annotations_amodal/*.json"))
+        gt_fpaths = list(self.gt_fpath.glob("*/per_sweep_annotations_amodal/*.json"))
+
+        assert len(dt_fpaths) == len(gt_fpaths)
+        data = defaultdict(list)
+        cls_inst_map = defaultdict(int)
+
+        with Pool(os.cpu_count()) as p:
+            accum = p.map(self.accumulate, gt_fpaths)
+        for frame_stats, frame_cls_to_inst in accum:
+            for cls_name, cls_stats in frame_stats.items():
+                data[cls_name].append(cls_stats)
+            for cls_name, num_inst in frame_cls_to_inst.items():
+                cls_inst_map[cls_name] += num_inst
+
+        data = defaultdict(np.ndarray, {k: np.vstack(v) for k, v in data.items()})
+
+        summary = self.summarize(data, cls_inst_map)
+        summary = (
+            pd.DataFrame.from_dict(summary, orient="index", columns=["AP", "ATE", "ASE", "AOE", "WDCS"])
+            .round(self.dt_cfg.significant_digits)
+            .sort_index()
+        )
+        summary.index = summary.index.str.capitalize()
+
+        summary.loc[""] = ["", "", "", "", ""]
+        summary.loc["Means"] = summary.iloc[:-1].mean().round(self.dt_cfg.significant_digits)
+        return summary
+
+    def accumulate(self, gt_fpath: Path) -> Tuple[np.ndarray, np.ndarray]:
+        """Accumulate the statistics for each LiDAR frame.
+
+        Args:
+            gt_fpath: The ground truth file path.
+
+        Returns:
+            The aggregated data used for summarization.
+        """
+        log_id = gt_fpath.parents[1].stem
+        logger.info("log_id = %s", log_id)
+        ts = gt_fpath.stem.split("_")[-1]
+        dt_fpath = self.dt_fpath / f"{log_id}/per_sweep_annotations_amodal/" f"tracked_object_labels_{ts}.json"
+
+        dts = np.array(read_label(dt_fpath))
+        gts = np.array(read_label(gt_fpath))
+        gt_clss = np.array([gt.label_class for gt in gts])
+
+        class_data = defaultdict(list)
+        class_to_ninst = defaultdict(int)
+        for gt_cls in np.unique(gt_clss):
+            dt_filtered = filter_instances(dts, gt_cls)
+            gt_filtered = filter_instances(gts, gt_cls)
+
+            logger.info("%d detections" % dt_filtered.shape[0])
+            logger.info("%d ground truth" % gt_filtered.shape[0])
+            if dt_filtered.shape[0] > 0:
+                error_types = self.assign(dt_filtered, gt_filtered)
+                class_data[gt_cls] = error_types
+
+            class_to_ninst[gt_cls] = gt_filtered.shape[0]
+        return class_data, class_to_ninst
+
+    def assign(self, dts: np.ndarray, gts: np.ndarray) -> List:
```

Comment:
let's add `List[specific_type]` to be more specific. But isn't the return type `np.ndarray`? Have we run mypy locally to make sure we catch any type mismatches?

Reply:
Haha, I was wondering how you caught all of these so quickly. Is this not included in the pre-commit hooks?

Reply:
actually caught these few manually, but `mypy` is quite useful